clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -302,27 +302,20 @@ private long sendProducerData(long now) {
     }
 
     private boolean maybeSendTransactionalRequest(long now) {
-        if (transactionManager.isCompletingTransaction() &&
```

Comment:
@apurvam I think this is probably a more likely cause of the test failure. This is a classic check-and-act scenario. It could very well be that `isCompletingTransaction()` returns false at this point, but becomes true by the time we call `nextRequestHandler` below. I inserted an artificial delay before the call to `nextRequestHandler` and was able to reproduce the test failure. To fix the issue, I pushed the safety check for incomplete batches into `nextRequestHandler`.